File: swift/Miolingo/Sources/MiolingoCore/VocabFunctions.swift
import Foundation

// =====================================================================
// VocabFunctions — ported verbatim (in behaviour) from spec/VocabFunctions.wl
// (recovered from src/vocab.py). Pure; no IO except the enrich oracle, which
// is passed in.
// =====================================================================

// --- derived identity + logical clock ---------------------------------
func vsNewId(_ entries: [VocabEntry]) -> Int { 1 + (entries.map(\.id).max() ?? 0) }
func vsNextSeq(_ entries: [VocabEntry]) -> Int { 1 + (entries.map(\.lastSeq).max() ?? 0) }

func emptyToNull(_ x: String?) -> String? {
    guard let x, !x.isEmpty else { return nil }
    return x
}
func coalesce(_ old: String?, _ new: String?) -> String? {
    if old == nil || old == "" { return emptyToNull(new) }
    return old
}

// --- _normalise + validate_single_word (vocab.py:31, 50) --------------
// _TRIM_PUNCT: ASCII + curly quotes, guillemets, dashes, ellipsis.
private let vsTrimChars: Set<Character> = {
    var s = Set(".,;:!?\"'-")
    for cp in [0x201C, 0x201D, 0x2018, 0x2019, 0xAB, 0xBB, 0x2014, 0x2013, 0x2026] {
        s.insert(Character(UnicodeScalar(cp)!))
    }
    return s
}()

/// Strip surrounding (never inner) trim-punctuation. Returns (display, key).
func normalise(_ word: String) -> (display: String, key: String) {
    var t = Substring(word.trimmingCharacters(in: .whitespacesAndNewlines))
    while let f = t.first, vsTrimChars.contains(f) { t = t.dropFirst() }
    while let l = t.last, vsTrimChars.contains(l) { t = t.dropLast() }
    let display = String(t)
    return (display, display.lowercased())
}

/// validate_single_word: returns (display, key) or nil.
public func validateWord(_ word: String) -> (display: String, key: String)? {
    if word.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return nil }
    let (display, key) = normalise(word)
    if key.isEmpty { return nil }                                   // only punctuation
    if key.rangeOfCharacter(from: .whitespacesAndNewlines) != nil { return nil } // not single word
    if key.count > 100 { return nil }                              // too long
    return (display, key)
}

// --- addEntry[entries, w] : capture_vocab_entry upsert (vocab.py:106) --
public func addEntry(_ entries: [VocabEntry], _ w: Capture) -> [VocabEntry] {
    guard let (display, key) = validateWord(w.word) else { return entries }
    if let idx = entries.firstIndex(where: { $0.word == key }) {
        var out = entries
        var m = out[idx]
        m.timesSeen += 1
        m.lastSeq = vsNextSeq(entries)                              // advance logical clock
        m.translation   = coalesce(m.translation, w.translation)
        m.ipa           = coalesce(m.ipa, w.ipa)
        m.sourceName    = coalesce(m.sourceName, w.sourceName)
        m.url           = coalesce(m.url, w.url)
        m.contextBefore = coalesce(m.contextBefore, w.contextBefore)
        m.contextLine   = coalesce(m.contextLine, w.contextLine)
        m.contextAfter  = coalesce(m.contextAfter, w.contextAfter)
        out[idx] = m
        return out
    } else {
        let seq = vsNextSeq(entries)
        let e = VocabEntry(
            id: vsNewId(entries), word: key, displayWord: display,
            translation: emptyToNull(w.translation), ipa: emptyToNull(w.ipa),
            sourceName: emptyToNull(w.sourceName), url: emptyToNull(w.url),
            contextBefore: emptyToNull(w.contextBefore),
            contextLine: emptyToNull(w.contextLine),
            contextAfter: emptyToNull(w.contextAfter),
            timesSeen: 1, firstSeq: seq, lastSeq: seq, notes: nil)
        return entries + [e]
    }
}

public func addEntry(_ entries: [VocabEntry], _ word: String) -> [VocabEntry] {
    addEntry(entries, Capture(word: word))
}

// --- deleteFrom[entries, id] : delete_vocab_entry (vocab.py:301) -------
public func deleteFrom(_ entries: [VocabEntry], _ id: Int) -> [VocabEntry] {
    entries.filter { $0.id != id }
}

// --- updateNotesIn (vocab.py:314) -------------------------------------
public func updateNotesIn(_ entries: [VocabEntry], id: Int, notes: String?) -> [VocabEntry] {
    entries.map { e in
        guard e.id == id else { return e }
        var m = e; m.notes = emptyToNull(notes); return m
    }
}

// --- updateEntry[entries, editingRow[id], fields] (vocab.py:343) -------
public let vsEditableFields: Set<String> = [
    "display_word", "translation", "ipa", "source_name",
    "url", "context_before", "context_line", "context_after"]

/// Apply an edit. Rejects unknown keys; rejects a display_word whose key would
/// change; "" → nil. On any rejection returns the list unchanged.
public func updateEntry(_ entries: [VocabEntry], id: Int, fields: [String: String]) -> [VocabEntry] {
    if !Set(fields.keys).isSubset(of: vsEditableFields) { return entries }
    guard let idx = entries.firstIndex(where: { $0.id == id }) else { return entries }
    let row = entries[idx]
    if let dw = fields["display_word"], normalise(dw).key != row.word { return entries }
    var out = entries
    var m = out[idx]
    for (k, v) in fields {
        let nv = emptyToNull(v)
        switch k {
        case "display_word":   if let nv { m.displayWord = nv }   // display kept; key unchanged
        case "translation":    m.translation = nv
        case "ipa":            m.ipa = nv
        case "source_name":    m.sourceName = nv
        case "url":            m.url = nv
        case "context_before": m.contextBefore = nv
        case "context_line":   m.contextLine = nv
        case "context_after":  m.contextAfter = nv
        default: break
        }
    }
    out[idx] = m
    return out
}

// --- autofillFields (vocab.py:411) — only-empty / never-overwrite -----
/// Compute the fill fields (translation/ipa) for entry `id` using the enrich
/// oracle and the borrowed language pair. Returns the fields to set (maybe empty).
public func autofillFields(_ entries: [VocabEntry], id: Int, lang: LangPair,
                           oracle: EnrichOracle) -> [String: String] {
    guard let row = entries.first(where: { $0.id == id }) else { return [:] }
    guard let fill = oracle.enrich(word: row.displayWord, source: lang.source, target: lang.target)
    else { return [:] }
    var out: [String: String] = [:]
    if (row.translation == nil || row.translation == ""), let t = emptyToNull(fill.translation) {
        out["translation"] = t
    }
    if (row.ipa == nil || row.ipa == ""), let i = emptyToNull(fill.ipa) {
        out["ipa"] = i
    }
    return out
}

// --- sort + filter (list_vocab order map + search; vocab.py:195) ------
public func sortEntries(_ entries: [VocabEntry], _ order: VocabSort) -> [VocabEntry] {
    switch order {
    case .alpha:  return entries.sorted { $0.word < $1.word }
    case .recent: return entries.sorted { $0.lastSeq > $1.lastSeq }
    case .oldest: return entries.sorted { $0.firstSeq < $1.firstSeq }
    }
}

func filterMatch(_ e: VocabEntry, _ filter: String?) -> Bool {
    guard let q = filter else { return true }
    let needle = q.lowercased()
    if needle.isEmpty { return true }
    if e.displayWord.lowercased().contains(needle) { return true }
    if (e.translation ?? "").lowercased().contains(needle) { return true }
    return false
}

public func applyFilter(_ entries: [VocabEntry], _ filter: String?) -> [VocabEntry] {
    entries.filter { filterMatch($0, filter) }
}

// --- practiseList (vocab.py:644) — shape to practice phrases -----------
public func practiseList(_ entries: [VocabEntry], filter: String?) -> [Phrase] {
    applyFilter(entries, filter).map {
        Phrase(text: $0.displayWord,
               translation: $0.translation ?? "",
               ipa: $0.ipa ?? "")
    }
}

File: web/app/src/domain/vocabFunctions.ts
// =====================================================================
// VocabFunctions — ported verbatim (in behaviour) from spec/VocabFunctions.wl
// (recovered from src/vocab.py). Pure; the enrich oracle's async fetch lives
// outside — only the fill policy (never-overwrite) is modelled here.
// =====================================================================

import type { Capture, Phrase, VocabEntry, VocabSort } from './types.js';

// --- derived identity + logical clock ---------------------------------
export function vsNewId(entries: readonly VocabEntry[]): number {
  return 1 + entries.reduce((m, e) => Math.max(m, e.id), 0);
}

export function vsNextSeq(entries: readonly VocabEntry[]): number {
  return 1 + entries.reduce((m, e) => Math.max(m, e.lastSeq), 0);
}

export function emptyToNull(x: string | null | undefined): string | null {
  return x != null && x !== '' ? x : null;
}

function coalesce(old: string | null, next: string | null | undefined): string | null {
  return old == null || old === '' ? emptyToNull(next) : old;
}

// --- _normalise + validate_single_word (vocab.py:31, 50) --------------
// _TRIM_PUNCT: ASCII + curly quotes, guillemets, dashes, ellipsis.
const TRIM_CHARS = new Set([
  ...'.,;:!?"\'-',
  '“',
  '”',
  '‘',
  '’',
  '«',
  '»',
  '—',
  '–',
  '…',
]);

/** Strip surrounding (never inner) trim-punctuation. Returns {display, key}. */
export function normaliseWord(word: string): { display: string; key: string } {
  let t = Array.from(word.trim());
  while (t.length > 0 && TRIM_CHARS.has(t[0]!)) t = t.slice(1);
  while (t.length > 0 && TRIM_CHARS.has(t[t.length - 1]!)) t = t.slice(0, -1);
  const display = t.join('');
  return { display, key: display.toLowerCase() };
}

/** validate_single_word: {display, key} or null. */
export function validateWord(word: string): { display: string; key: string } | null {
  if (word.trim() === '') return null;
  const { display, key } = normaliseWord(word);
  if (key === '') return null; // only punctuation
  if (/\s/.test(key)) return null; // not a single word
  if (Array.from(key).length > 100) return null; // too long
  return { display, key };
}

// --- addEntry[entries, w] : capture_vocab_entry upsert (vocab.py:106) --
export function addEntry(entries: readonly VocabEntry[], w: Capture | string): VocabEntry[] {
  const cap: Capture = typeof w === 'string' ? { word: w } : w;
  const valid = validateWord(cap.word);
  if (valid === null) return [...entries];
  const { display, key } = valid;
  const idx = entries.findIndex((e) => e.word === key);
  if (idx >= 0) {
    const m = entries[idx]!;
    const updated: VocabEntry = {
      ...m,
      timesSeen: m.timesSeen + 1,
      lastSeq: vsNextSeq(entries), // advance the logical clock
      translation: coalesce(m.translation, cap.translation),
      ipa: coalesce(m.ipa, cap.ipa),
      sourceName: coalesce(m.sourceName, cap.sourceName),
      url: coalesce(m.url, cap.url),
      contextBefore: coalesce(m.contextBefore, cap.contextBefore),
      contextLine: coalesce(m.contextLine, cap.contextLine),
      contextAfter: coalesce(m.contextAfter, cap.contextAfter),
    };
    return entries.map((e, i) => (i === idx ? updated : e));
  }
  const seq = vsNextSeq(entries);
  const entry: VocabEntry = {
    id: vsNewId(entries),
    word: key,
    displayWord: display,
    translation: emptyToNull(cap.translation),
    ipa: emptyToNull(cap.ipa),
    sourceName: emptyToNull(cap.sourceName),
    url: emptyToNull(cap.url),
    contextBefore: emptyToNull(cap.contextBefore),
    contextLine: emptyToNull(cap.contextLine),
    contextAfter: emptyToNull(cap.contextAfter),
    timesSeen: 1,
    firstSeq: seq,
    lastSeq: seq,
    notes: null,
  };
  return [...entries, entry];
}

// --- deleteFrom[entries, id] : delete_vocab_entry (vocab.py:301) -------
export function deleteFrom(entries: readonly VocabEntry[], id: number): VocabEntry[] {
  return entries.filter((e) => e.id !== id);
}

// --- updateNotesIn (vocab.py:314) -------------------------------------
export function updateNotesIn(
  entries: readonly VocabEntry[],
  id: number,
  notes: string | null,
): VocabEntry[] {
  return entries.map((e) => (e.id === id ? { ...e, notes: emptyToNull(notes) } : e));
}

// --- updateEntry[entries, editingRow[id], fields] (vocab.py:343) -------
export const EDITABLE_FIELDS: ReadonlySet<string> = new Set([
  'display_word',
  'translation',
  'ipa',
  'source_name',
  'url',
  'context_before',
  'context_line',
  'context_after',
]);

/**
 * Apply an edit. Rejects unknown keys; rejects a display_word whose key would
 * change; "" → null. On any rejection returns the list unchanged.
 */
export function updateEntry(
  entries: readonly VocabEntry[],
  id: number,
  fields: Readonly<Record<string, string>>,
): VocabEntry[] {
  if (!Object.keys(fields).every((k) => EDITABLE_FIELDS.has(k))) return [...entries];
  const idx = entries.findIndex((e) => e.id === id);
  if (idx < 0) return [...entries];
  const row = entries[idx]!;
  const dw = fields['display_word'];
  if (dw !== undefined && normaliseWord(dw).key !== row.word) return [...entries];
  let m = { ...row };
  for (const [k, v] of Object.entries(fields)) {
    const nv = emptyToNull(v);
    switch (k) {
      case 'display_word':
        if (nv !== null) m = { ...m, displayWord: nv }; // display kept; key unchanged
        break;
      case 'translation':
        m = { ...m, translation: nv };
        break;
      case 'ipa':
        m = { ...m, ipa: nv };
        break;
      case 'source_name':
        m = { ...m, sourceName: nv };
        break;
      case 'url':
        m = { ...m, url: nv };
        break;
      case 'context_before':
        m = { ...m, contextBefore: nv };
        break;
      case 'context_line':
        m = { ...m, contextLine: nv };
        break;
      case 'context_after':
        m = { ...m, contextAfter: nv };
        break;
    }
  }
  return entries.map((e, i) => (i === idx ? m : e));
}

// --- autofill (vocab.py:411) — only-empty / never-overwrite ------------
/** What an enrich oracle returns for a word (translation and/or IPA). */
export interface EnrichFill {
  readonly translation?: string | null;
  readonly ipa?: string | null;
}

/**
 * Pure autofill policy: given the oracle's fill for entry `id`, compute the
 * fields to set — only ones currently empty; never overwrite. (The async
 * oracle fetch itself happens in the AppModel, borrowing the language pair
 * fresh via langRead.)
 */
export function autofillFields(
  entries: readonly VocabEntry[],
  id: number,
  fill: EnrichFill | null,
): Record<string, string> {
  const row = entries.find((e) => e.id === id);
  if (row === undefined || fill === null) return {};
  const out: Record<string, string> = {};
  const t = emptyToNull(fill.translation);
  if ((row.translation == null || row.translation === '') && t !== null) out['translation'] = t;
  const i = emptyToNull(fill.ipa);
  if ((row.ipa == null || row.ipa === '') && i !== null) out['ipa'] = i;
  return out;
}

// --- sort + filter (list_vocab order map + search; vocab.py:195) ------
export function sortEntries(entries: readonly VocabEntry[], order: VocabSort): VocabEntry[] {
  const out = [...entries];
  switch (order) {
    case 'alpha':
      return out.sort((a, b) => (a.word < b.word ? -1 : a.word > b.word ? 1 : 0));
    case 'recent':
      return out.sort((a, b) => b.lastSeq - a.lastSeq);
    case 'oldest':
      return out.sort((a, b) => a.firstSeq - b.firstSeq);
  }
}

function filterMatch(e: VocabEntry, filter: string | null): boolean {
  if (filter == null) return true;
  const needle = filter.toLowerCase();
  if (needle === '') return true;
  if (e.displayWord.toLowerCase().includes(needle)) return true;
  return (e.translation ?? '').toLowerCase().includes(needle);
}

export function applyFilter(entries: readonly VocabEntry[], filter: string | null): VocabEntry[] {
  return entries.filter((e) => filterMatch(e, filter));
}

// --- practiseList (vocab.py:644) — shape to practice phrases -----------
export function practiseList(entries: readonly VocabEntry[], filter: string | null): Phrase[] {
  return applyFilter(entries, filter).map((e) => ({
    text: e.displayWord,
    translation: e.translation ?? '',
    ipa: e.ipa ?? '',
  }));
}
